### DISCUSSION TITLE
Artifacts on surfaces that touch

### DISCUSSION THREAD

**Discussion by Dzejkop** [bevyengine/bevy#3747] [Q&A]
I'm seeing some weird artifacts on surfaces that are touching. Depending on the angle of the camera the surfaces will render through one another.

Here's a gif
![mcskin](https://user-images.githubusercontent.com/5088647/150659039-4592c5e8-64e7-468a-95b7-640b1345f0ea.gif)

How can I fix this?

**Comment by nicopap** [comment]
It would be helpful if you provide more details. Could you answer those questions:
* How do you generate the mesh, is this a gltf file you import? Or just some cube meshes you generate in your code?
* Is the texture semi-transparent?

This looks like the faces of your cuboids are facing inwards, so that they do not hide the things that are behind them. This may be [related to this](https://github.com/bevyengine/bevy/issues/3030). I can't recommend a solution, maybe [double-faced culling mod could help, once implemented](https://github.com/bevyengine/bevy/issues/3729)

In the case of a gltf file, I found that sometimes the meshes would generate with the surfaces facing the wrong way, and I never figured out the proper fix.

Hint, this is not Z-fighting (unlike the title of your post would suggest)

**Reply by Dzejkop** [reply]
These meshes are imported from GLTF. But I think the faces are oriented right, if they weren't I should also see the same effect with just a single mesh, right? But I'm only seeing this with multiple meshes.

Maybe helpful, when I imported these meshes to Godot I had to set `Depth Draw Mode` to `Always` on the material to prevent this.

**Reply by nicopap** [reply]
Yeah, I had similar issues with a gltf model. I think bevy relies on vertex ordering to compute surface orientation when loading gltf. It seems blender manages to screw up vertex ordering enough when exporting that it confused bevy.

What happens is that the faces of your model are facing inward, so you can see "into" the model. It is especially visible with lighting, it just reflects light as if the model was carved out.

Again, no idea how to fix this :/

**Comment by superdump** [comment]
This almost looks to me like the faces of the meshes are not being drawn using a depth buffer and so the order of the draw calls is making one or the other pop on top based on the view space z position of the model origin. Are the faces all part of one single mesh, or are they multiple meshes? How are you rendering these? Using a plain PbrBundle as loaded by the bevy gltf loader? Modifying those somehow?

**Reply by superdump** [reply]
Is the mesh marked as needing alpha blending? If you make it opaque does the problem go away? My guess is that it will. Transparent meshes (those mesh entities with AlphaMode::Blend) are sorted back to front based on the view z of the origin of the mesh and transparent meshes don’t write to the depth buffer. Opaque do so even if they’re sorted ‘incorrectly’ the fragments that are further away than something else that was already drawn will not be shaded.

**Reply by Dzejkop** [reply]
They're separate meshes, each spawned with a PbrBundle. They share the same material and on material I have `alpha_mode` set to `AlphaMode::Blend`. I need them to support transparency though.

**Reply by superdump** [reply]
Can you change the alpha mode when they actually become transparent and if they have a colour with an alpha of 1.0 then you set the alpha mode to opaque?

**Reply by Dzejkop** [reply]
The meshes will actually be transparent by default, so I can't do that.

**Reply by superdump** [reply]
If the materials had an alpha value non-trivially less than 1.0 then perhaps Weighted-Blended Order-Independent Transparency (WBOIT) would help your case: https://github.com/bevyengine/bevy/issues/2223 But if your alpha values approach 1.0 then nvidia say that the results diverge from the ground truth. I don’t know in what way though.

bevy doesn’t support WBOIT yet though.
